.github/workflows/build.yml

```diff
@@ -49,63 +49,45 @@ jobs:
               })
             }
             console.log("Clear completed")
-      - name: Login to Docker Hub
-        uses: docker/login-action@v3
+      - name: Install dependencies
+        uses: ./.github/actions/install_dependencies
         with:
-          username: ${{ secrets.DOCKERHUB_USERNAME }}
-          password: ${{ secrets.DOCKERHUB_TOKEN }}
-      - name: Install required packages
-        run: |
-          sudo apt-get update
-          sudo apt install -y build-essential libssl-dev zlib1g-dev libbz2-dev libreadline-dev libsqlite3-dev curl \
-              git libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev zip unzip \
-              libpython3.11
-      - name: Install Python
-        uses: gabrielfalcao/pyenv-action@v18
-        with:
-          default: 3.11.4
+          docker-username: ${{ secrets.DOCKERHUB_USERNAME }}
+          docker-token: ${{ secrets.DOCKERHUB_TOKEN }}
+      #NB: We restore/save cache manually so that we save the cache even if the build fails
       - name: Load m2 repository cache # Manually caching .m2 repo as the setup-java caching isn't falling back to older caches
         id: cached-m2-repo
-        uses: actions/cache@v4
+        uses: actions/cache/restore@v4
         with:
           path: ~/.m2/repository
           key: maven-${{ hashFiles('**/pom.xml') }}
           restore-keys: |
             maven-
       - name: Load m2 build cache
         id: cached-m2-build
-        uses: actions/cache@v4
+        uses: actions/cache/restore@v4
         with:
           path: ~/.m2/build-cache
           key: maven-build-cache-${{ hashFiles('**/pom.xml') }}
           restore-keys: |
             maven-build-cache-
-      - name: Install Poetry
-        uses: snok/install-poetry@v1
+      - name: Load docker build cache
+        id: cached-docker-build
+        uses: actions/cache/restore@v4
+        with:
+          path: ~/.docker/cache
+          key: docker-cache-${{ hashFiles('**/Dockerfile') }}
+          restore-keys: |
+            docker-cache-
+      #NB: Not saving poetry cache on failure in case it's a failure caused by an in-flight python package release
       - name: Poetry cache
         id: cached-poetry
         uses: actions/cache@v4
         with:
           path: ~/.cache/pypoetry
-          key: poetry-cache-${{ hashFiles('**/pom.xml') }}
+          key: poetry-cache-${{ hashFiles('**/pyproject.toml') }}
```

Comment:
I: Use pyproject.toml so the cache is updated when the python dependencies change, even if the POMs have not changed.